e2e/controls.py

```diff
@@ -0,0 +1,116 @@
+"""Facade classes for working with Shiny inputs/outputs in Playwright"""
+
+# pyright: reportUnknownMemberType=false
+
+from playwright.sync_api import Page, expect, Locator
+
+
+class SimpleInput:
+    def __init__(self, page: Page, selector: str):
+        self.loc = page.locator(selector)
+
+    @property
+    def expect(self):
+        return expect(self.loc)
+
+
+class TextInput(SimpleInput):
+    def __init__(self, page: Page, inputId: str):
+        super().__init__(page, f"input#{inputId}[type=text].shiny-bound-input")
+
+
+class TextAreaInput(SimpleInput):
+    def __init__(self, page: Page, inputId: str):
+        super().__init__(page, f"textarea#{inputId}.shiny-bound-input")
+
+
+class NumericInput(SimpleInput):
+    def __init__(self, page: Page, inputId: str):
+        super().__init__(page, f"input#{inputId}[type=number].shiny-bound-input")
+
+
+class CheckboxInput(SimpleInput):
+    def __init__(self, page: Page, inputId: str):
+        super().__init__(page, f"input#{inputId}[type=checkbox].shiny-bound-input")
+
+
+class DateInput:
+    def __init__(self, page: Page, inputId: str):
+        self.loc = page.locator(f"#{inputId}")
+
+    @property
+    def expect(self):
+        return expect(self.loc.locator("input"))
+
+
+class DateRangeInput:
+    def __init__(self, page: Page, inputId: str):
+        self.loc = page.locator(f"#{inputId}")
+
+    @property
+    def expect_start(self):
+        return expect(self.loc.locator("input:first-of-type"))
+
+    @property
+    def expect_end(self):
+        return expect(self.loc.locator("input:last-of-type"))
+
+
+class SliderInput:
+    def __init__(self, page: Page, inputId: str):
+        self.loc = page.locator(f"#{inputId}.js-range-slider").locator("xpath=..")
+
+    def move_slider(self, fraction: float) -> None:
```

Comment:
Perhaps this could be done in a less brittle way by calling `setValue` on the input binding (and/or working with ion range slider API more directly)?

Reply:
I specifically wanted our end-to-end tests to work "like a user". It might also be useful to have a way to do `setValue`, but the purpose for that would be to test that slider's `setValue` works, whereas `move_slider` tests that slider's UI works.